Refactor eos_cli_config - block 09 - TG

titom73 opened the pull request:
## Change Summary

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes)
- [ ] Documentation content changes
- [ ] Other (please describe):

## Related Issue(s)

N/A

## Component(s) name

`arista.avd.eos_cli_config_gen`

__Eos Templates__: `vlan-interfaces.j2`

## Proposed changes

- [x] Use `arista.avd.defined` test where possible
- [x] Use `arista.avd.default` filter where possible
- [x] Shorten long EOS CLI
- [x] Remove default values
- [x] Update Molecule to increase coverage
- [ ] Rename tasks to use collection paths: `ansible.builtin.template`
- [x] Move to concatenation using `~` instead of `+` to automatically convert to `string`
- [ ] Use `x in []` instead of multiple `if / elif / else` blocks
- [ ] Use `| join` instead of oneline `for` loop in documentation

## Submitted files

- [x] virtual-source-nat.j2
- [x] vlan-interfaces.j2
- [x] vlan-internal-order.j2
- [x] vlans.j2
- [ ] vmtracer-sessions.j2
- [ ] vrf-instances.j2
- [ ] vxlan-interface.j2

## How to test

Molecule should not introduce change. Only spaces should be updated

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://www.avd.sh/docs/contributing/) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly
- [x] All new and existing tests passed ([`pre-commit`](https://www.avd.sh/docs/installation/development/#python-virtual-environment), `make linting` and `make sanity-lint`).

titom73 committed:
Refactor eos_cli_config vlan-interfaces

titom73 committed:
Refactor vlan-interfaces as per peer-review

- Remove unecessary if arista.avd.defined
- Be explicit for MTU settings
- Be explicit for VRRP priority
- Fix indentation

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,166 @@
-{%         endif %}
-{%         if vlan_interfaces[vlan_interface].ipv6_enable is defined and vlan_interfaces[vlan_interface].ipv6_enable == true %}
+{%     endif %}
+{%    for ip_helper in vlan_interfaces[vlan_interface].ip_helpers | arista.avd.natural_sort %}
```
Missing one space for indentation between line 42 and line 49

titom73 replied:
Fixed with  07d5adb6

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,166 @@
-{%             endif %}
-{%         if vlan_interfaces[vlan_interface].ipv6_nd_ra_disabled is defined and vlan_interfaces[vlan_interface].ipv6_nd_ra_disabled == true %}
-   ipv6 nd ra disabled
 {%         endif %}
```
fix indentation

titom73 replied:
Fixed with  07d5adb6

ClausHolbechArista replied:
not seeing the change here. Line 60 is still indented much more than line 58.

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,166 @@
-{%           if vlan_interfaces[vlan_interface].ospf_authentication == "simple" %}
+{%     endif %}
+{%     if vlan_interfaces[vlan_interface].ospf_authentication is arista.avd.defined %}
+{%       if vlan_interfaces[vlan_interface].ospf_authentication == "simple" %}
```
fix indentation in lines 107, 109, 111

titom73 replied:
Fixed with  07d5adb6

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,166 @@
-{%         if vlan_interfaces[vlan_interface].ospf_message_digest_keys is defined and vlan_interfaces[vlan_interface].ospf_message_digest_keys is not none %}
-{%           for ospf_message_digest_key in vlan_interfaces[vlan_interface].ospf_message_digest_keys | arista.avd.natural_sort %}
+{%     endif %}
+{%     if vlan_interfaces[vlan_interface].ospf_message_digest_keys is arista.avd.defined %}
```
this check is included in arista.avd.natural_sort, so you can remove this.

titom73 replied:
Fixed with  07d5adb6

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,166 @@
-{%            if vlan_interfaces[vlan_interface].pim.ipv4.sparse_mode is defined and vlan_interfaces[vlan_interface].pim.ipv4.sparse_mode == true %}
+{%       endfor %}
+{%     endif %}
+{%     if vlan_interfaces[vlan_interface].pim is arista.avd.defined %}
```
with out latest findings, we don't need this check, since arista.avd.defined used below will check the full chain.

titom73 replied:
Fixed with  07d5adb6

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,166 @@
-{%           if vlan_interfaces[vlan_interface].vrrp.virtual_router is defined and vlan_interfaces[vlan_interface].vrrp.virtual_router is not none %}
-{%             if vlan_interfaces[vlan_interface].vrrp.priority is defined and vlan_interfaces[vlan_interface].vrrp.priority is not none and vlan_interfaces[vlan_interface].vrrp.priority > 1 %}
+{%     endif %}
+{%     if vlan_interfaces[vlan_interface].vrrp is arista.avd.defined %}
```
with out latest findings, we don't need this check, since arista.avd.defined used below will check the full chain.

titom73 replied:
Fixed with  07d5adb6

ClausHolbechArista reviewed (changes requested):
A few more optimizations are possible and a few indentation issues.

titom73 committed:
Fix indentation as per review

titom73 committed:
Fix indentation & avd filter as per review

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,160 @@
-{%         endif %}
-{%         if vlan_interfaces[vlan_interface].isis_enable is defined %}
+{%     endif %}
+{%     if vlan_interfaces[vlan_interface].isis_enable is defined %}
```
missed this before. Use arista.avd.defined :)

titom73 replied:
Fixed in commit 1b8cf2d

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,160 @@
-{%             endif %}
-{%             if vlan_interfaces[vlan_interface].vrrp.preempt_delay_minimum is defined and vlan_interfaces[vlan_interface].vrrp.preempt_delay_minimum is not none %}
+{%         endif %}
+{%        if vlan_interfaces[vlan_interface].vrrp.preempt_delay_minimum is arista.avd.defined %}
```
missing one leading space.

titom73 replied:
Fixed in commit 1b8cf2d

ClausHolbechArista reviewed (approved):
LGTM

titom73 committed:
Fix multiple whitespaces rendering in vlan-interfaces.j2

titom73 committed:
Update CI artifacts

titom73 committed:
Refactor Virtual Source NAT template

titom73 committed:
Refactor VLAN Internal Order template

titom73 committed:
Refactor VLANs template

titom73 committed:
Increase Molecule coverage for vlans, vlan-internal-order, virtual-source-nat

titom73 commented:
New commits for following files:

- virtual-source-nat
- vlans: change code to also report `state: active` and not only `state: suspend`
- vlan-internal-order

Increase Molecule coverage for templates above.

onurgashi commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,166 @@
+{%     endif %}
+{%     if vlan_interfaces[vlan_interface].multicast.ipv4.source_route_export.enabled is arista.avd.defined(true) %}
+{%         if vlan_interfaces[vlan_interface].multicast.ipv4.source_route_export.administrative_distance is arista.avd.defined %}
+   multicast ipv4 source route export {{ vlan_interfaces[vlan_interface].multicast.ipv4.source_route_export.administrative_distance | arista.avd.default('') }}
```
Do we need an else, if we set a default of '' - nothing. Because the if condition on line 83 fulfils both cases right?

As if you set nothing it will still print no admin distance?

onurgashi commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,166 @@
-{%         endif %}
-{%     endfor %}
-{% endif %}
+{%     elif vlan_interfaces[vlan_interface].ip_attached_host_route_export is arista.avd.defined %}
```
Do we need this elif, if we set the default in the line 161 as empty using the new filter?

ClausHolbechArista replied:
using default to "" would render a blank space instead (which was what we had before), so this change is correct.

titom73 committed:
Remove unused default filter for multicast and host-route

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/virtual-source-nat.j2:
```diff
@@ -1,5 +1,5 @@
 {# eos - virtual source-nat - introduced in 4.21 #}
-{% if virtual_source_nat_vrfs is defined and virtual_source_nat_vrfs is not none %}
+{% if virtual_source_nat_vrfs is arista.avd.defined %}
```
This check is included in arista.avd.natural_sort, so you can remove this.

titom73 replied:
Indeed, It is because we start the section with a `!` and it helps to not put complexity in the loop

```
{% if virtual_source_nat_vrfs is arista.avd.defined %}
!
{%     for vrf in virtual_source_nat_vrfs | arista.avd.natural_sort %}
```

ClausHolbechArista replied:
Got it. Good point

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/virtual-source-nat.j2:
```diff
@@ -1,5 +1,5 @@
+{% if virtual_source_nat_vrfs is arista.avd.defined %}
 !
 {%     for vrf in virtual_source_nat_vrfs | arista.avd.natural_sort %}
 ip address virtual source-nat vrf {{ vrf }} address {{ virtual_source_nat_vrfs[vrf].ip_address }}
```
Missing check for ip_address.

titom73 replied:
Fixed with commit d42782a527b53acfb04121f714f34954f98cf806

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,166 @@
+{%             if vlan_interfaces[vlan_interface].ipv6_nd_prefixes[prefix].valid_lifetime is arista.avd.defined %}
+{%                 set ipv6_nd_prefix_cli = ipv6_nd_prefix_cli ~ " " ~ vlan_interfaces[vlan_interface].ipv6_nd_prefixes[prefix].valid_lifetime %}
+{%             endif %}
+{%             if vlan_interfaces[vlan_interface].ipv6_nd_prefixes[prefix].preferred_lifetime is arista.avd.defined %}
```
`preferred_lifetime` is only valid in CLI together with `valid_lifetime`, so I suggest to move this section inside the `if` for `valid_lifetime`

titom73 replied:
Fixed with commit d42782a527b53acfb04121f714f34954f98cf806

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-interfaces.j2:
```diff
@@ -1,156 +1,166 @@
-{%           for ospf_message_digest_key in vlan_interfaces[vlan_interface].ospf_message_digest_keys | arista.avd.natural_sort %}
+{%     endif %}
+{%     for ospf_message_digest_key in vlan_interfaces[vlan_interface].ospf_message_digest_keys | arista.avd.natural_sort %}
    ip ospf message-digest-key {{ ospf_message_digest_key }} {{ vlan_interfaces[vlan_interface].ospf_message_digest_keys[ospf_message_digest_key].hash_algorithm }} 7 {{ vlan_interfaces[vlan_interface].ospf_message_digest_keys[ospf_message_digest_key].key }}
```
please add checks for `hash_algoritm` and `key`

titom73 replied:
Fixed with commit d42782a527b53acfb04121f714f34954f98cf806

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlan-internal-order.j2:
```diff
@@ -1,5 +1,5 @@
 {# Internal VLAN allocation policy  #}
-{% if vlan_internal_order is defined and vlan_internal_order is not none %}
+{% if vlan_internal_order is arista.avd.defined %}
```
change to check for `vlan_internal_order.allocation` *and* `vlan_internal_order.range.beginning` *and* `vlan_internal_order.range.ending`

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlans.j2:
```diff
@@ -1,16 +1,12 @@
+{% for vlan in vlans | arista.avd.natural_sort %}
 !
 vlan {{ vlan }}
    name {{ vlans[vlan].name }}
```
add check for name

titom73 replied:
Fixed with commit d42782a527b53acfb04121f714f34954f98cf806

ClausHolbechArista commented on ansible_collections/arista/avd/roles/eos_cli_config_gen/templates/eos/vlans.j2:
```diff
@@ -1,16 +1,12 @@
 {%     endfor %}
-{% endif %}
\ No newline at end of file
+{% endfor %}
```
missing newline at the end of the file

titom73 replied:
Fixed with commit d42782a527b53acfb04121f714f34954f98cf806

ClausHolbechArista reviewed (changes requested):
Looks good, but I have added a few requests for changes.

titom73 committed:
refactor(eos_cli_config_gen): Refactor based on code review

Update following templates:
- virtual-source-nat.j2
- vlan-interfaces.j2
- vlan-internal-order.j2
- vlans.j2

titom73 committed:
Merge remote-tracking branch 'upstream/refactor/eos-cli-config-gen' into refactor/eos-cli-config-gen-vlan-interfaces

ClausHolbechArista reviewed (approved):
LGTM
